glider_dac/glider_emails.py

```diff
@@ -62,3 +81,104 @@ def get_erddap_catalog_url():
     }
     url = 'http://%(host)s/erddap/metadata/iso19115/xml/' % args
     return url
+
+def glider_deployment_check(data_type=None, completed=True, force=False,
+                            deployment_dir=None, username=None):
+    """
+    """
+    cs = CheckSuite()
+    cs.load_all_available_checkers()
+    with app.app_context():
+        if data_type is not None:
+            is_delayed_mode = data_type == 'delayed'
+            if is_delayed_mode:
+                q_dict = {"delayed_mode": True,
+                          "completed": completed}
+            else:
+                q_dict = {"$or": [{"delayed_mode": False},
+                                  {"delayed_mode": {"$exists": False}}],
+                          "completed": completed}
+
+            if not force:
+                q_dict["compliance_check_passed"] = {"$ne": True}
+
+        # TODO: combine username/deployment cases?
+        if username:
+            q_dict = {"username": username}
+        # a particular deployment has been specified
+        elif deployment_dir:
+            q_dict = {"deployment_dir": deployment_dir}
+        else:
+            q_dict = {}
+
+        agg_pipeline = [{"$match": q_dict},
+                        {"$group": {"_id": "$user_id",
+                           "deployments": {"$push":
+                                {"_id": "$_id",
+                                 "name": "$name",
+                                 "deployment_dir": "$deployment_dir"} } }
+                        }]
+        # if force is enabled, re-check the datasets no matter what
+
+        # is this syntax still used?  if the first fn call fails, use the
+        # second set of results
+        try:
+            agg_result_set = db.deployments.aggregate(agg_pipeline)['result']
+        except:
+            agg_result_set = db.deployments.aggregate(agg_pipeline, cursor={})
```

Comment:
Can you explain what the agg pipeline is about? This is getting all the deployments in the DAC and checking every one of them? What triggers this to run?

Reply:
Certain queries in MongoDB can only run via the "aggregation framework", such as queries that require grouping.  The syntax to call it also changed in MongoDB 2 to 3, so that's why there's a try catch block here to handle both cases.

Reply:
This does not necessarily get all the deployments.  There are a number of options to limit to a particular user or deployment, to process delayed mode, and whether or not to reprocess compliance checks which have already completed (force) or which can be invoked calling the email CLI options.